internal/overlord/servstate/manager.go

```diff
@@ -53,6 +53,76 @@ func (m *ServiceManager) reloadSetup() error {
 	return nil
 }
 
+// FlattenedSetup returns the flattened setup as a single layer in YAML format.
+func (m *ServiceManager) FlattenedSetup() ([]byte, error) {
+	releaseSetup, err := m.acquireSetup()
+	if err != nil {
+		return nil, err
+	}
+	defer releaseSetup()
+
+	return m.flattened.AsYAML()
+}
+
+// MergeLayer merges the given layer YAML into the dynamic layers, returning
+// the new layer's "order" (won't increase if adding another dynamic layer).
+func (m *ServiceManager) MergeLayer(layerYAML []byte) (int, error) {
+	layer, err := setup.ParseLayer(0, "", layerYAML)
+	if err != nil {
+		return 0, err
+	}
+
+	releaseSetup, err := m.acquireSetup()
+	if err != nil {
+		return 0, err
+	}
+	defer releaseSetup()
+
+	var last *setup.Layer
+	layers := m.setup.Layers
+	if len(layers) > 0 {
+		last = layers[len(layers)-1]
+	}
+
+	var newOrder int
+	var newSetup *setup.Setup
+	var newFlattened *setup.Layer
+	if last != nil && last.IsDynamic() {
+		// Last layer is dynamic, merge new layer into existing dynamic layer
+		dynamicSetup := &setup.Setup{Layers: []*setup.Layer{last, layer}}
+		dynamicFlattened, err := dynamicSetup.Flatten()
+		if err != nil {
+			return 0, err
+		}
+		dynamicFlattened.Order = last.Order
+		newOrder = dynamicFlattened.Order
+		newSetup = &setup.Setup{Layers: append(layers[:len(layers)-1], dynamicFlattened)}
+		newFlattened, err = newSetup.Flatten()
+		if err != nil {
+			return 0, err
+		}
+	} else if last != nil {
```

Comment:
Let's just simplify this to
```
if last != nil && last.IsDynamic() {
		// Last layer is dynamic, merge new layer into existing dynamic layer
		dynamicSetup := &setup.Setup{Layers: []*setup.Layer{last, layer}}
		dynamicFlattened, err := dynamicSetup.Flatten()
		if err != nil {
			return 0, err
		}
		dynamicFlattened.Order = last.Order
		newOrder = dynamicFlattened.Order
		newSetup = &setup.Setup{Layers: append(layers[:len(layers)-1], dynamicFlattened)}
		newFlattened, err = newSetup.Flatten()
		if err != nil {
			return 0, err
		}
	} else {
		// Last layer is not dynamic, add new dynamic layer
		layer.Order = 1
		if last != nil {
			layer.Order = last.Order + 1
		}
		newOrder = layer.Order
		newSetup = &setup.Setup{Layers: append(layers, layer)}
		newFlattened, err = newSetup.Flatten()
		if err != nil {
			return 0, err
		}
	}
```

Reply:
Heh, I was on the fence about this and actually had it the other way at first, but then I changed it because "flat is better than nested". But I think you're right, this way is slightly simple/shorter, so changed it back. :-)